File: AgentBI/src/api/image_generation.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Request, Response, status

from AgentBI.src.api.dependencies import get_chat_repository
from AgentBI.src.schemas.image_generation_schema import (
    CodexOAuthStartResponse,
    CodexOAuthStatus,
    GeneratedImage,
    ImageGenerationRequest,
)
from AgentBI.src.services.image_generation.codex_oauth import CodexOAuthError
from AgentBI.src.services.image_generation.lite_adapter import ImageGenerationError


router = APIRouter(prefix="/image-generation", tags=["image-generation"])


def _service(request: Request):
    service = getattr(request.app.state, "image_generation_service", None)
    if not service:
        raise HTTPException(status_code=503, detail="图片生成服务尚未初始化")
    return service


def _oauth(request: Request):
    manager = getattr(request.app.state, "codex_image_oauth", None)
    if not manager:
        raise HTTPException(status_code=503, detail="Codex OAuth 服务尚未初始化")
    return manager


@router.post("/generate", response_model=GeneratedImage)
async def generate_image(request: Request, payload: ImageGenerationRequest):
    repository = get_chat_repository(request)
    provider = None
    if payload.provider_id:
        provider = repository.get_provider(payload.provider_id)
        if not provider:
            raise HTTPException(status_code=404, detail="模型提供商不存在")
    try:
        return await _service(request).generate(
            user_id=payload.user_id,
            scope_id=payload.scope_id or "direct",
            prompt=payload.prompt,
            aspect_ratio=payload.aspect_ratio,
            provider=provider,
            reference_image_data_url=payload.reference_image_data_url,
        )
    except (ImageGenerationError, CodexOAuthError) as error:
        raise HTTPException(status_code=400, detail=str(error)) from error


@router.get("/codex/status", response_model=CodexOAuthStatus)
def codex_status(request: Request):
    return _oauth(request).status()


@router.post("/codex/connect", response_model=CodexOAuthStartResponse)
async def connect_codex(request: Request):
    try:
        return await _oauth(request).start()
    except CodexOAuthError as error:
        raise HTTPException(status_code=409, detail=str(error)) from error


@router.delete("/codex/connection", status_code=status.HTTP_204_NO_CONTENT)
def disconnect_codex(request: Request):
    _oauth(request).disconnect()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: Agent-vue/src/api/image-generation.ts
import type {
  CodexImageOAuthStart,
  CodexImageOAuthStatus,
  GeneratedImage,
  GenerateImagePayload,
} from './image-generation-types'

const BASE = '/api/image-generation'

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${BASE}${path}`, init)
  if (!response.ok) {
    const payload = await response.json().catch(() => null)
    throw new Error(payload?.detail || `图片生成请求失败 (${response.status})`)
  }
  if (response.status === 204) return undefined as T
  return response.json() as Promise<T>
}

export const generateImage = (payload: GenerateImagePayload) =>
  request<GeneratedImage>('/generate', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })

export const getCodexImageStatus = () =>
  request<CodexImageOAuthStatus>('/codex/status')

export const connectCodexImage = () =>
  request<CodexImageOAuthStart>('/codex/connect', { method: 'POST' })

export const disconnectCodexImage = () =>
  request<void>('/codex/connection', { method: 'DELETE' })

